darkcluster/src/main/java/com/linkedin/darkcluster/impl/ConstantQpsDarkClusterStrategy.java

```diff
@@ -0,0 +1,132 @@
+/*
+   Copyright (c) 2021 LinkedIn Corp.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package com.linkedin.darkcluster.impl;
+
+import com.linkedin.common.callback.Callback;
+import com.linkedin.common.util.None;
+import com.linkedin.darkcluster.util.GuaranteedRateLimiter;
+import javax.annotation.Nonnull;
+
+import com.linkedin.common.util.Notifier;
+import com.linkedin.d2.DarkClusterConfig;
+import com.linkedin.d2.balancer.ServiceUnavailableException;
+import com.linkedin.d2.balancer.util.ClusterInfoProvider;
+import com.linkedin.darkcluster.api.BaseDarkClusterDispatcher;
+import com.linkedin.darkcluster.api.DarkClusterStrategy;
+import com.linkedin.r2.message.RequestContext;
+import com.linkedin.r2.message.rest.RestRequest;
+
+/**
+ * ConstantQpsDarkClusterStrategy figures out how many dark requests to send. The high level goal of this strategy is to
+ * keep the incoming QPS per dark host constant.
+ *
+ * It uses the {@link ClusterInfoProvider} to determine the number of instances in both the source and target cluster,
+ * and uses that to calculate the number of requests to send in order to make the QPS per dark host constant and equal
+ * to a specified value, assuming all hosts in the source cluster send traffic.
+ */
+public class ConstantQpsDarkClusterStrategy implements DarkClusterStrategy {
+  private final String _originalClusterName;
+  private final String _darkClusterName;
+  private final Integer _darkClusterPerHostInboundTargetRate;
+  private final BaseDarkClusterDispatcher _baseDarkClusterDispatcher;
+  private final Notifier _notifier;
+  private final ClusterInfoProvider _clusterInfoProvider;
+  private final GuaranteedRateLimiter _rateLimiter;
+
+  public ConstantQpsDarkClusterStrategy(@Nonnull String originalClusterName, @Nonnull String darkClusterName,
+      @Nonnull Integer darkClusterPerHostInboundTargetRate, @Nonnull BaseDarkClusterDispatcher baseDarkClusterDispatcher,
```

Comment:
Minor: Do you think darkClusterPerHostQPS is a more informative name than inboundTargetRate?

Reply:
Yeah, I like darkClusterPerHostQps better